File: 프로그래머스/2/131127. 할인 행사/할인 행사.py
from collections import defaultdict

def solution(want, number, discount):
    answer = 0
    
    want_dict = defaultdict(int)
    curr_dict = defaultdict(int)
    
    for i in range(len(want)):
        want_dict[want[i]] = number[i]
        
    for i in range(10):
        curr_dict[discount[i]] += 1
    
    if check_contain(want_dict, curr_dict): answer += 1
    
    ## 슬라이딩 윈도우
    for i in range(10, len(discount)):
        ## 한 칸 밀기
        curr_dict[discount[i-10]] -= 1
        curr_dict[discount[i]] += 1
        
        if check_contain(want_dict, curr_dict): answer += 1
    
    return answer

def check_contain(want_dict, curr_dict):
    for key, value in want_dict.items():
        if key not in curr_dict or curr_dict[key] < value:
            return False
        
    return True

File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        
        Map<String, Integer> wantMap = new HashMap<>();
        Map<String, Integer> currMap = new HashMap<>();
        
        for(int i=0; i<want.length; ++i){
            wantMap.put(want[i], number[i]);
        }
        
        for(int i=0; i<10; ++i){
            currMap.put(discount[i], currMap.getOrDefault(discount[i], 0)+1);
        }
        
        if(checkContains(wantMap, currMap)) answer++;
        
        // 슬라이딩 윈도우
        for(int i=10; i<discount.length; ++i){
            // 한 칸 밀기
            currMap.put(discount[i-10], currMap.get(discount[i-10])-1); 
            currMap.put(discount[i], currMap.getOrDefault(discount[i], 0)+1);
            
            if(checkContains(wantMap, currMap)) answer++;
        }
        
        return answer;
    }
    
    public boolean checkContains(Map<String, Integer> wantMap, Map<String, Integer> currMap){
        for(Map.Entry<String, Integer> entry: wantMap.entrySet()){
            if(!currMap.containsKey(entry.getKey()) || currMap.get(entry.getKey()) < entry.getValue()) return false;
        }
        
        return true;
    }
}